lib/membrane/video_compositor/wgpu/wgpu_adapter.ex

```diff
@@ -1,165 +1,79 @@
 
   @doc """
-  Uploads a frame to the compositor.
-
-  If all videos have provided input frames with a current enough pts, this will also render and return a composed frame.
+  Initialize the new part of the compositor
   """
-  @spec process_frame(wgpu_state(), id(), frame_with_pts()) ::
-          :ok | {:ok, frame_with_pts()}
-  def process_frame(state, id, {frame, pts}) do
-    case Native.process_frame(state, id, frame, pts) do
-      :ok ->
-        :ok
-
-      {:ok, frame} ->
-        {:ok, frame}
+  @spec init() :: native_state()
+  def init() do
+    case Native.init() do
+      {:ok, state} ->
+        state
 
       {:error, reason} ->
-        raise "Error while uploading/composing frame, reason: #{inspect(reason)}"
+        raise "Error while initializing the compositor, reason: #{inspect(reason)}"
     end
   end
 
-  @doc """
-  Forcibly renders the composed frame, even if we are still waiting for some frames to arrive
-  """
-  @spec force_render(state :: wgpu_state()) ::
-          {:ok, merged_frames :: frame_with_pts()} | {:error, error()}
-
-  def force_render(state) do
-    case Native.force_render(state) do
-      {:ok, frame} -> {:ok, frame}
-      {:error, reason} -> raise "Error while force rendering, reason: #{inspect(reason)}"
-    end
+  # Gets the wgpu context necessary for initializing transformation modules and layout modules.
+  #
+  # Safety
+  # It's vital that this struct is used according to the safety sections in the docs for the rust
+  # mirror of the struct this function returns
+  # (`membrane_video_compositor_common::elixir_transfer::StructElixirPacket::<WgpuContext>`)
+  @spec wgpu_ctx(native_state()) :: wgpu_ctx()
+  defp wgpu_ctx(state) do
+    Native.wgpu_ctx(state)
   end
 
   @doc """
-  Add an input video with the given numerical `id`.
-
-  Provided `id` should be unique within all previous ones. If a video with a given id already exists in the compositor, this will raise.
+  This function takes a list of transformation modules, initializes them and registers them in
+  the compositor so that they're available to use in the scene.
   """
-  @spec add_video(
-          state :: wgpu_state(),
-          id :: id(),
-          stream_format :: Membrane.RawVideo.t(),
-          placement :: RustStructs.BaseVideoPlacement.t(),
-          transformations :: VideoTransformations.t()
-        ) :: :ok | {:error, error()}
-  def add_video(state, id, stream_format, placement, transformations) do
-    {:ok, rust_stream_format} = RustStructs.RawVideo.from_membrane_raw_video(stream_format)
-
-    case Native.add_video(state, id, rust_stream_format, placement, transformations) do
-      :ok -> :ok
-      {:error, reason} -> raise "Error while adding a video, reason: #{inspect(reason)}"
-    end
+  @spec register_transformations(
+          native_state(),
+          list(Transformation.transformation_module())
+        ) :: :ok
+  def register_transformations(state, transformations) do
```

Comment:
perhaps `add_transformations` or something similar would be better than `register_transformations` since the function not only "registers" the transformations in the native part, but it also initializes them?

Reply:
That's very fair